peps/pep-0749.rst

```diff
@@ -0,0 +1,741 @@
+
+.. code:: pycon
+
+  >>> fields(D)[0].type
+  ForwardRef('undefined')
+
+We could instead add deferred evaluation for the field type, similar to that outlined
+above for type alias values.
+
+Accessing ``.type`` might throw an error:
+
+.. code:: pycon
+
+  >>> @dataclass
+  ... class D:
+  ...     x: undefined
+  ...     
+  >>> field = fields(D)[0]
+  >>> field.type
+  Traceback (most recent call last):
+    File "<python-input-4>", line 1, in <module>
+      field.type
+    File ".../dataclasses.py", line 308, in type
+      annos = self._annotate(annotations.Format.VALUE)
+    File "<python-input-2>", line 3, in __annotate__
+      x: undefined
+        ^^^^^^^^^
+  NameError: name 'undefined' is not defined
+
+But users could use ``annotations.call_evaluate_function`` to get the type in other formats:
+
+.. code:: pycon
+
+  >>> annotations.call_evaluate_function(field.evaluate_type, annotations.Format.SOURCE)
+  'undefined'
+  >>> annotations.call_evaluate_function(field.evaluate_type, annotations.Format.FORWARDREF)
+  ForwardRef('undefined')
+
+Other variations are possible. For example, we could leave the ``type`` attribute unchanged,
+and only add the ``evaluate_type`` method. This avoids unpleasant surprises where accessing
+``.type`` may throw an exception.
+
+Acknowledgments
+===============
+
+First of all, I thank Larry Hastings for writing :pep:`649`. This PEP modifies some of his
+initial decisions, but the overall design is still his.
+
+I thank Carl Meyer and Alex Waygood for feedback on early drafts of this PEP.
+
+Appendix
+========
+
+Which expressions can be stringified?
+-------------------------------------
+
+:pep:`649` acknowledges that the stringifier cannot handle all expressions. Now that we
+have a draft implementation, we can be more precise about the expressions that can and
+cannot be handled. Below is a list of all expressions in the Python AST that can and
+cannot be recovered by the stringifier. The full list should probably not be added to
+the documentation, but creating it is a useful exercise.
+
+First, the stringifier of course cannot recover any information that is not present in
+the compiled code, including comments, whitespace, parenthesization, and operations that
+get simplified by the AST optimizer.
+
+Second, the stringifier can intercept almost all operations that involve names looked
+up in some scope, but it cannot interject operations that operate fully on constants.
+As a corollary, this also means it is not safe to request SOURCE format on untrusted
+code: Python is powerful enough that it is possible to achieve arbitrary code execution
+even with no access to any globals or builtins. For example:
+
+.. code-block:: pycon
+
+  >>> def f(x: (1).__class__.__base__.__subclasses__()[-1].__init__.__builtins__["print"]("Hello world")): pass
+  ... 
+  >>> annotations.get_annotations(f, format=annotations.Format.SOURCE)
+  Hello world
+  {'x': 'None'}
```

Comment:
you'll get PRs for the next 5 years trying to code-golf this example down

Reply:
Lol, adding this does introduce some maintenance risk. It also relies on the most recently created subclass of `object` having an `__init__` method, which is probably usually true but it may not always hold.